### DISCUSSION TITLE
Sensor SimulationBody/BodyFixture prevents other bodies from becoming at rest

### DISCUSSION THREAD

**Discussion by NordelDev** [dyn4j/dyn4j#185] [Q&A] [answered]
Hi William,

I have the following "problem".

I have several `SimulationBody`s which are continously checked (via. StepListener). Based on these `SimulationBody`s I define an `AABB` (via union), which I then use to define the `BodyFixture` for a "surrounding" Box around all `SimulationBody` of interest.

This `BodyFixture` of the surrounding `SimulationBody` is updated every `TimeStep`. As suggested on your homepage I remove the `BodyFixture` of the previous `TimeStep` and add a new one (based on the updated AABB sizes) to the surrounding `SimulationBody`.

```
@Override
	public void end(TimeStep step, PhysicsWorld<SimulationBody, ?> world)
	{
		SimulationBody simulationBodyAABB = this.workstation.getSimulationBodyAABB();
		
		// Loop over all SimulationBody
		AABB workstationAABB = null;
		AABB resourceBaseAABB;
		for (SimulationBody simulationBody : this.includedSimulationBodies)
		{
			// First BodyFixture is the "BaseFootprint"
			resourceBaseAABB = simulationBody.getFixture(0).getShape().createAABB();
			simulationBody.getFixture(0).getShape().computeAABB(simulationBody.getTransform(), resourceBaseAABB);
			
			if(workstationAABB == null)
			{
				workstationAABB = resourceBaseAABB;
			}
			else
			{
				workstationAABB = workstationAABB.union(resourceBaseAABB);
			}
		}
		

		/*
		 * Remove the current BodyFixture
		 */
		simulationBodyAABB.removeAllFixtures();
		
		
		/*
		 * Create a new BodyFixture, which represents the current dimensions of the AABB
		 */
		double width = workstationAABB.getWidth();
		double height = workstationAABB.getHeight();
		
		BodyFixture aabbBodyFixture = new BodyFixture(Geometry.createRectangle(width, height));
		BodyFixtureData bfd = new BodyFixtureData();
		bfd.setShapePrimitive(ShapePrimitive.RECTANGLE);
		bfd.setIsToRender(true);
		aabbBodyFixture.setUserData(bfd);
		aabbBodyFixture.setSensor(true);
		simulationBodyAABB.addFixture(aabbBodyFixture);
			
		
		/*
		 * Translate the SimulationBody to its new position
		 */
		Vector2 aabbCenter = workstationAABB.getCenter();
 		simulationBodyAABB.translateToOrigin();
		simulationBodyAABB.translate(aabbCenter);


                /*
		 * Translate the SimulationBody to its new position
		 */
		Vector2 aabbCenter = workstationAABB.getCenter();
 		simulationBodyAABB.translateToOrigin();
		simulationBodyAABB.translate(aabbCenter);
		
		
		/*
		 * Update the SimulationBody of the Workstation
		 */
		this.workstation.setSimulationBodyAABB(simulationBodyAABB);
}
```

Initially I define the surrounding `SimulationBody` the following way:

```
/**
 * Updating WorkstationAABB
 */
				
	SimulationBody workstationAABB = new SimulationBody();
	BodyFixture aabbBodyFixture = new BodyFixture(Geometry.createRectangle(0.0001, 0.0001));	// Very small
	BodyFixtureData bfd = new BodyFixtureData();
	bfd.setShapePrimitive(ShapePrimitive.RECTANGLE);
	bfd.setIsToRender(true);
	aabbBodyFixture.setUserData(bfd);
	aabbBodyFixture.setSensor(true);
	workstationAABB.addFixture(aabbBodyFixture);
	workstationAABB.setMass(MassType.INFINITE);
	currentWorkstation.setSimulationBodyAABB(workstationAABB);
	this.world.addBody(workstationAABB);
				
	WorkstationAabbUpdater workstationAabbUpdater = new WorkstationAabbUpdater(currentWorkstation, workstationSimulationBodys);
	this.world.addStepListener(workstationAabbUpdater);
```

In that case `workstationSimulationBodys` are the `SimulationBody`s, which should be considered for the `AABB` of a "Workstation".

Basically this works quite well. But I am facing the problem, that the further code execution depends on the status check, if all `SimulationBody`, which should be surrounded, are `atRest`.

If I don't include the `SimulationBody` with surrounding boundingbox, my code executes further as wished after all `SimulationBody`s are at rest.

I have the feeling, that even if the surrounding `SimulationBody` is **set as sensor**, it activates somehow the other `SimulationBody`s which results in the problem, that my code is trapped in some kind of endless loop since the detected `SimulationBody`s are evertime activated again.

Video with surrounding `SimulationBody`:

![With_surrounding](https://user-images.githubusercontent.com/34217758/109315817-3ca10d00-784b-11eb-8d50-9ff503c2e3d4.gif)

Video without surrounding `SimulationBody`:

![Without_surrounding](https://user-images.githubusercontent.com/34217758/109315822-3dd23a00-784b-11eb-8239-c5c5d1da665b.gif)

Here are then some calculations executed, which are only visible in the console :)

---

I need these surrounding AABBs because I run some calculations in the background (e.g. MinimumDistance between 2 AABBs)

---

Any idea is highly appreciated :)

Thank you in advance.

**Comment by wnbittle** [comment]
Is there a reason you are adding the union `AABB` as a `Body` to the simulation?

Either way, I think it's because you are adding/removing the `AABB` every frame.  When you remove a `Fixture` from a `Body`, the world checks for any `ContactConstraints` involving that `Fixture`.  If it finds one, it removes it and sets the body to `atRest` = `false`.  Sensors generate `ContactConstraints` even though they are not solved.  Depending on your use case, you could implement a `BroadphaseCollisionDataFilter` or a listener to stop the detection system from creating `ContactConstraints` and causing this issue.  

Do you need the collision data from the sensor?

**Reply by NordelDev** [reply]
This has some visualization purposes, so that I don't need another list with Bodies or Shapes which should be rendered.

Basically that I can use already existing methods :)

I need the surrounding AABB for some distance calculations betweens specific groups of `SimulationBody`. And because of the above mentioned reason of rendering I "translate" the AABB into a `SimulationBody`. Right now I don't need collision data but maybe for a later use.

**Comment by wnbittle** [comment, marked as answer]
I released a new version with a fix for this issue.  Give it a go when you get a chance and let me know if that fixed the issue.

**Reply by NordelDev** [reply]
Thank you very much. I will test it today or tomorrow and provide you some feedback.

And I am sorry for the late reply.

**Reply by NordelDev** [reply]
Good morning.

With the new version you uploaded everything works like I a charm :) Thank you very much for your quick action.
